src/go/k8s/cmd/syncclusterconfig/sync.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+// Package syncclusterconfig contains a re-implementation of the original
+// redpanda helm chart's post-upgrade and post-install jobs in go via direct
+// connections to the AdminAPI rather than templating out rpk commands.
+//
+// The original scripts can be found here:
+// - https://github.com/redpanda-data/helm-charts/blob/2bf9dd6f09966189cb00a9b5beab49d3ae57b0e1/charts/redpanda/post_install_upgrade_job.go#L109-L164
+// - https://github.com/redpanda-data/helm-charts/blob/2bf9dd6f09966189cb00a9b5beab49d3ae57b0e1/charts/redpanda/post_upgrade_job.go#L82-L141
+package syncclusterconfig
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/redpanda-data/common-go/rpadmin"
+	rpkadminapi "github.com/redpanda-data/redpanda/src/go/rpk/pkg/adminapi"
+	rpkconfig "github.com/redpanda-data/redpanda/src/go/rpk/pkg/config"
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	licenseEnvvar = "REDPANDA_LICENSE"
+)
+
+func Command() *cobra.Command {
+	var redpandaYAMLPath string
+	var bootstrapYAMLPath string
+
+	cmd := &cobra.Command{
+		Use: "sync-cluster-config --bootstrap-yaml file [--redpanda-yaml file]",
+		Long: fmt.Sprintf(`sync-cluster-config patches a cluster's configuration with values from the provided bootstrap.yaml.
+If present and not empty, the $%s environment variable will be set as the cluster's license.
+`, licenseEnvvar),
+		SilenceUsage: true, // Don't show --help when errors are returned from RunE
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx := cmd.Context()
+			logger := log.FromContext(ctx)
+
+			client, err := adminAPIFromRPKConfig(redpandaYAMLPath)
+			if err != nil {
+				return err
+			}
+
+			if license := os.Getenv(licenseEnvvar); license == "" {
+				logger.Info(fmt.Sprintf("$%s not set. Skipping license setting...", licenseEnvvar))
+			} else {
+				logger.Info(fmt.Sprintf("$%s set. Setting license...", licenseEnvvar))
+				if err := client.SetLicense(cmd.Context(), strings.NewReader(license)); err != nil {
+					return err
+				}
+			}
+
+			clusterConfig, err := loadBoostrapYAML(bootstrapYAMLPath)
+			if err != nil {
+				return err
+			}
+
+			// NB: remove must be an empty slice NOT nil.
+			result, err := client.PatchClusterConfig(ctx, clusterConfig, []string{})
+			if err != nil {
+				return err
+			}
+
+			logger.Info("Updated cluster configuration", "config_version", result.ConfigVersion)
+
+			// NB: It's unclear why this endpoint is being hit. It's preserved
+			// as a historical artifact.
+			if err := client.RestartService(ctx, "schema-registry"); err != nil {
```

Comment:
I don't remember why this restart is needed. The PR which introduced this change have 0 context. I can only guess now that cluster which is upgraded to from no SASL to cluster with SASL enabled might have problem with Schema Registry. It would be good to create follow up to create test case for such upgrade and remove this schema registry restart.

https://github.com/redpanda-data/helm-charts/pull/846

https://redpandadata.atlassian.net/wiki/spaces/CORE/pages/326500353/Schema+Registry+Restart

https://github.com/redpanda-data/redpanda/pull/9631

Reply:
I'll file an issue to circle back on this one and leave it in place for now unless you have any major objections?

Reply:
No objection I tried to provide context to the best of my ability.